ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java

```diff
@@ -417,65 +407,79 @@ private static boolean blockDescendsFromLatestFinalizedBlock(
-        "on_attestation: Attestations must be from the current or previous epoch");
+    if (!target.getEpoch().equals(previous_epoch) && !target.getEpoch().equals(current_epoch)) {
+      return AttestationProcessingResult.invalid(
+          "on_attestation: Attestations must be from the current or previous epoch");
+    }
+
+    if (!target.getEpoch().equals(compute_epoch_at_slot(attestation.getData().getSlot()))) {
+      return AttestationProcessingResult.invalid(
+          "on_attestation: Attestation slot must be within specified epoch");
+    }
 
-    checkArgument(
-        target.getEpoch().equals(compute_epoch_at_slot(attestation.getData().getSlot())),
-        "on_attestation: Attestation slot must be within specified epoch");
+    // Attestations can only affect the fork choice of subsequent slots.
+    // Delay consideration in the fork choice until their slot is in the past.
+    if (get_current_slot(store).compareTo(attestation.getData().getSlot()) <= 0) {
+      return AttestationProcessingResult.FAILED_NOT_FROM_PAST;
+    }
 
-    checkArgument(
-        store.getBlockRoots().contains(target.getRoot()),
-        "on_attestation: Attestations target must be for a known block. If a target block is unknown, delay consideration until the block is found");
+    if (!store.getBlockRoots().contains(target.getRoot())) {
+      // Attestations target must be for a known block. If a target block is unknown, delay
+      // consideration until the block is found
+      return AttestationProcessingResult.FAILED_UNKNOWN_TARGET;
+    }
 
     // Attestations cannot be from future epochs. If they are, delay consideration until the epoch
     // arrives
-    BeaconState targetRootState = store.getBlockState(target.getRoot());
-    checkArgument(
-        get_current_slot(store).compareTo(compute_start_slot_at_epoch(target.getEpoch())) >= 0,
-        "on_attestation: Attestations cannot be from the future epochs");
-
-    checkArgument(
-        store.getBlockRoots().contains(attestation.getData().getBeacon_block_root()),
-        "on_attestation: Attestations must be for a known block. If block is unknown, delay consideration until the block is found");
-
-    checkArgument(
-        store
-                .getBlock(attestation.getData().getBeacon_block_root())
-                .getSlot()
-                .compareTo(attestation.getData().getSlot())
-            <= 0,
-        "on_attestation: Attestations must not be for blocks in the future. If not, the attestation should not be considered");
+    if (get_current_slot(store).compareTo(target.getEpochSlot()) < 0) {
+      return AttestationProcessingResult.FAILED_FUTURE_EPOCH;
+    }
+
+    if (!store.getBlockRoots().contains(attestation.getData().getBeacon_block_root())) {
+      // Attestations must be for a known block. If block is unknown, delay consideration until the
+      // block is found
+      return AttestationProcessingResult.FAILED_UNKNOWN_TARGET;
```

Comment:
Should this be renamed:
```suggestion
      return AttestationProcessingResult.FAILED_UNKNOWN_BLOCK;
```
?

Reply:
Yeah makes sense.